File: default/lib/outcome-inference/outcome_inference.py
#!/usr/bin/env python3
"""Outcome inference module for automatic handoff outcome detection.

This module analyzes session state to infer task outcomes:
- SUCCEEDED: All todos complete, no errors
- PARTIAL_PLUS: Most todos complete, minor issues
- PARTIAL_MINUS: Some progress, significant blockers
- FAILED: No meaningful progress or abandoned

Usage:
    from outcome_inference import infer_outcome
    outcome = infer_outcome(project_root)
"""

import json
import os
import re
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Tuple


# Outcome definitions
OUTCOME_SUCCEEDED = "SUCCEEDED"
OUTCOME_PARTIAL_PLUS = "PARTIAL_PLUS"
OUTCOME_PARTIAL_MINUS = "PARTIAL_MINUS"
OUTCOME_FAILED = "FAILED"
OUTCOME_UNKNOWN = "UNKNOWN"


def get_project_root() -> Path:
    """Get the project root directory."""
    if os.environ.get("CLAUDE_PROJECT_DIR"):
        return Path(os.environ["CLAUDE_PROJECT_DIR"])
    return Path.cwd()


def sanitize_session_id(session_id: str) -> str:
    """Sanitize session ID for safe use in filenames."""
    sanitized = re.sub(r'[^a-zA-Z0-9_-]', '', session_id)
    if not sanitized:
        return "unknown"
    return sanitized[:64]


def analyze_todos(todos: List[Dict]) -> Tuple[int, int, int, int]:
    """Analyze todo list and return counts.

    Returns:
        Tuple of (total, completed, in_progress, pending)
    """
    if not todos:
        return (0, 0, 0, 0)

    total = len(todos)
    completed = sum(1 for t in todos if t.get("status") == "completed")
    in_progress = sum(1 for t in todos if t.get("status") == "in_progress")
    pending = sum(1 for t in todos if t.get("status") == "pending")

    return (total, completed, in_progress, pending)


def infer_outcome_from_todos(
    total: int,
    completed: int,
    in_progress: int,
    pending: int
) -> Tuple[str, str]:
    """Infer outcome from todo counts.

    Returns:
        Tuple of (outcome, reason)
    """
    if total == 0:
        return (OUTCOME_FAILED, "No todos tracked - unable to measure completion")

    completion_ratio = completed / total

    if completed == total:
        return (OUTCOME_SUCCEEDED, f"All {total} tasks completed")

    if completion_ratio >= 0.8:
        remaining = total - completed
        return (OUTCOME_PARTIAL_PLUS, f"{completed}/{total} tasks done, {remaining} minor items remain")

    if completion_ratio >= 0.5:
        return (OUTCOME_PARTIAL_MINUS, f"{completed}/{total} tasks done, significant work remains")

    # <50% completion is FAILED - insufficient progress
    # Note: in_progress tasks without completion don't count toward progress
    return (OUTCOME_FAILED, f"Insufficient progress: only {completed}/{total} tasks completed ({int(completion_ratio * 100)}%)")


def infer_outcome(
    todos: Optional[List[Dict]] = None,
) -> Dict:
    """Infer session outcome from available state.

    Args:
        todos: Optional pre-parsed todos list

    Returns:
        Dict with 'outcome', 'reason', and 'confidence' keys
    """
    result = {
        "outcome": OUTCOME_UNKNOWN,
        "reason": "Unable to determine outcome",
        "confidence": "low",
        "sources": [],
    }

    if todos is not None:
        total, completed, in_progress, pending = analyze_todos(todos)
        # Always infer outcome from todos - even if empty (total=0 -> FAILED)
        result["outcome"], result["reason"] = infer_outcome_from_todos(
            total, completed, in_progress, pending
        )
        result["sources"].append("todos")
        # Confidence based on completion clarity
        if total == 0:
            # Empty todos = high confidence it's a failure
            result["confidence"] = "high"
        elif completed == total or completed == 0:
            result["confidence"] = "high"
        elif completed / total >= 0.8 or completed / total <= 0.2:
            result["confidence"] = "medium"
        else:
            result["confidence"] = "low"

    return result


def main():
    """CLI entry point for testing."""
    import sys

    # Read optional todos from stdin
    todos = None
    if not sys.stdin.isatty():
        try:
            input_data = json.loads(sys.stdin.read())
            todos = input_data.get("todos")
        except json.JSONDecodeError:
            pass

    result = infer_outcome(todos=todos)
    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: opencode/.opencode/plugin/outcome-inference.ts
import type { Plugin } from "@opencode-ai/plugin"
import { readState, writeState, getSessionId } from "./utils/state"
import { EVENTS } from "./utils/events"

/**
 * Ring Outcome Inference Plugin
 * Infers session outcome from todo state.
 *
 * Equivalent to: default/lib/outcome-inference/outcome_inference.py
 * Event: session.idle
 *
 * This plugin runs when a session goes idle and infers the outcome
 * based on todo completion status. The outcome is saved for
 * learning extraction to pick up.
 *
 * NOTE: outcome-inference and learning-extract both use session.idle.
 * learning-extract should run after outcome-inference completes.
 */

// Outcome constants
const OUTCOME = {
  SUCCEEDED: "SUCCEEDED",
  PARTIAL_PLUS: "PARTIAL_PLUS",
  PARTIAL_MINUS: "PARTIAL_MINUS",
  FAILED: "FAILED",
  UNKNOWN: "UNKNOWN",
} as const

type OutcomeType = (typeof OUTCOME)[keyof typeof OUTCOME]

interface Todo {
  content: string
  status: "pending" | "in_progress" | "completed"
}

interface OutcomeResult {
  outcome: OutcomeType
  reason: string
  confidence: "low" | "medium" | "high"
  sources: string[]
}

interface SessionOutcome {
  status: OutcomeType
  summary: string
  confidence: string
  inferred_at: string
  key_decisions: string[]
}

/**
 * Analyze todos and return counts.
 */
function analyzeTodos(todos: Todo[]): { total: number; completed: number; inProgress: number; pending: number } {
  if (!todos || todos.length === 0) {
    return { total: 0, completed: 0, inProgress: 0, pending: 0 }
  }

  return {
    total: todos.length,
    completed: todos.filter((t) => t.status === "completed").length,
    inProgress: todos.filter((t) => t.status === "in_progress").length,
    pending: todos.filter((t) => t.status === "pending").length,
  }
}

/**
 * Infer outcome from todo counts.
 */
function inferOutcomeFromTodos(
  total: number,
  completed: number,
  inProgress: number,
  pending: number
): { outcome: OutcomeType; reason: string } {
  if (total === 0) {
    return { outcome: OUTCOME.FAILED, reason: "No todos tracked - unable to measure completion" }
  }

  const completionRatio = completed / total

  if (completed === total) {
    return { outcome: OUTCOME.SUCCEEDED, reason: `All ${total} tasks completed` }
  }

  if (completionRatio >= 0.8) {
    const remaining = total - completed
    return { outcome: OUTCOME.PARTIAL_PLUS, reason: `${completed}/${total} tasks done, ${remaining} minor items remain` }
  }

  if (completionRatio >= 0.5) {
    return { outcome: OUTCOME.PARTIAL_MINUS, reason: `${completed}/${total} tasks done, significant work remains` }
  }

  // <50% completion is FAILED
  return {
    outcome: OUTCOME.FAILED,
    reason: `Insufficient progress: only ${completed}/${total} tasks completed (${Math.floor(completionRatio * 100)}%)`,
  }
}

/**
 * Infer outcome from available state.
 */
function inferOutcome(todos: Todo[] | null): OutcomeResult {
  const result: OutcomeResult = {
    outcome: OUTCOME.UNKNOWN,
    reason: "Unable to determine outcome",
    confidence: "low",
    sources: [],
  }

  if (todos !== null) {
    const { total, completed, inProgress, pending } = analyzeTodos(todos)
    const inferred = inferOutcomeFromTodos(total, completed, inProgress, pending)

    result.outcome = inferred.outcome
    result.reason = inferred.reason
    result.sources.push("todos")

    // Confidence based on completion clarity
    if (total === 0) {
      result.confidence = "high" // Empty todos = high confidence it's a failure
    } else if (completed === total || completed === 0) {
      result.confidence = "high"
    } else if (completed / total >= 0.8 || completed / total <= 0.2) {
      result.confidence = "medium"
    } else {
      result.confidence = "low"
    }
  }

  return result
}

export const RingOutcomeInference: Plugin = async ({ directory }) => {
  const projectRoot = directory
  const sessionId = getSessionId()

  const IDEMPOTENCY_WINDOW_MS = 10_000

  return {
    event: async ({ event }) => {
      if (event.type !== EVENTS.SESSION_IDLE) {
        return
      }

      const alreadyRan = readState<{ timestamp: number }>(projectRoot, "outcome-inference-ran", sessionId)
      if (alreadyRan && Date.now() - alreadyRan.timestamp < IDEMPOTENCY_WINDOW_MS) {
        return
      }

      console.log("[Ring:INFO] Session idle - inferring outcome...")

      // Read todos state
      const todos = readState<Todo[]>(projectRoot, "todos-state", sessionId)

      // Infer outcome
      const result = inferOutcome(todos)

      // Save outcome for learning extraction
      const outcome: SessionOutcome = {
        status: result.outcome,
        summary: result.reason,
        confidence: result.confidence,
        inferred_at: new Date().toISOString(),
        key_decisions: [],
      }

      writeState(projectRoot, "session-outcome", outcome, sessionId)
      writeState(projectRoot, "outcome-inference-ran", { timestamp: Date.now() }, sessionId)

      console.log(
        `[Ring:INFO] Outcome inferred: ${result.outcome} (${result.confidence} confidence) - ${result.reason}`
      )
    },
  }
}

export default RingOutcomeInference
